### PR TITLE
crimson: futurized CyanStore's member functions and Collection

### PR DESCRIPTION
**Pull request by liu-chunmei**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by liu-chunmei** [comment]
@tchaikov I split the alien store pr to some certain features,  and still debug on the alien blue store, please review these first.

File: src/crimson/osd/osd.cc [hunk 1]
```diff
@@ -123,16 +123,19 @@ seastar::future<> OSD::mkfs(uuid_d osd_uuid, uuid_d cluster_fsid)
...
       cluster_fsid,
       superblock.osd_fsid);
-
-    meta_coll = make_unique<OSDMeta>(
-      store->create_new_collection(coll_t::meta()), store.get());
-    ceph::os::Transaction t;
-    meta_coll->create(t);
-    meta_coll->store_superblock(t, superblock);
-    return store->do_transaction(meta_coll->collection(), std::move(t));
+    return store->create_new_collection(coll_t::meta()).then([this] (auto ch) {
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc]: no need to embed this continuation. instead, we can just 1) put it into the next one 2) lift the next continuation. like
```c++
}).then([cluster_fsid, this] {
  return store->create_new_collection(coll_t::meta());
}).then([this](auto collection) {
  ceph::os::Transaction t;
  meta_coll = make_unique<OSDMeta>(collection , store.get());
  meta_coll->create(t);
  // ...
```

File: src/crimson/osd/osd.cc [hunk 2]
```diff
@@ -186,17 +189,18 @@ seastar::future<> OSD::start()
...
-    return meta_coll->load_superblock();
-  }).then([this](OSDSuperblock&& sb) {
-    superblock = std::move(sb);
-    return get_map(superblock.current_epoch);
-  }).then([this](cached_map_t&& map) {
-    shard_services.update_map(osdmap);
-    osdmap_gate.got_map(map->get_epoch());
-    osdmap = std::move(map);
-    return load_pgs();
+    return store->open_collection(coll_t::meta()).then([this](auto ch) {
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc]: ditto. please avoid embed the continuations if possible.

File: src/crimson/osd/osd.cc [hunk 3]
```diff
@@ -385,8 +389,8 @@ seastar::future<> OSD::stop()
 
 seastar::future<> OSD::load_pgs()
 {
-  return seastar::parallel_for_each(store->list_collections(),
-    [this](auto coll) {
+  return store->list_collections().then([this](auto colls){
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc]: might want to add a space before `{`.

File: src/crimson/osd/pg.cc [hunk 4]
```diff
@@ -192,9 +193,10 @@ void PG::init(
 
 seastar::future<> PG::read_state(ceph::os::FuturizedStore* store)
 {
-  coll_ref = store->open_collection(coll_t(pgid));
-  return PGMeta{store, pgid}.load().then(
-    [this, store](pg_info_t pg_info, PastIntervals past_intervals) {
+  return store->open_collection(coll_t(pgid)).then([this, store](auto ch){
+     coll_ref = ch;
+    return PGMeta{store, pgid}.load().then(
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: wrong indent.

Comment (tchaikov) [reply src/crimson/osd/pg.cc]: again, please flatten the continuation chain.

File: src/crimson/osd/pg_meta.cc [hunk 5]
```diff
@@ -31,8 +31,8 @@ namespace {
 }
 seastar::future<epoch_t> PGMeta::get_epoch()
 {
-  auto ch = store->open_collection(coll_t{pgid});
-  return store->omap_get_values(ch,
+  return store->open_collection(coll_t{pgid}).then ([this](auto ch){
```
Comment (tchaikov) [inline src/crimson/osd/pg_meta.cc]: remove the space after `then`.

File: src/crimson/osd/pg_meta.cc [hunk 6]
```diff
@@ -51,12 +51,13 @@ seastar::future<epoch_t> PGMeta::get_epoch()
...
       }
     });
+  });
 }
 
 seastar::future<pg_info_t, PastIntervals> PGMeta::load()
 {
-  auto ch = store->open_collection(coll_t{pgid});
-  return store->omap_get_values(ch,
+  return store->open_collection(coll_t{pgid}).then([this](auto ch){
```
Comment (tchaikov) [inline src/crimson/osd/pg_meta.cc]: add a space before `{`.

File: src/crimson/os/cyan_collection.h [hunk 7]
```diff
@@ -24,12 +25,8 @@ class Object;
  * ObjectStore users my get collection handles with open_collection() (or,
  * for bootstrapping a new collection, create_new_collection()).
  */
-struct Collection : public boost::intrusive_ref_counter<
-  Collection,
-  boost::thread_unsafe_counter>
-{
+struct Collection : public ObjectStore::CollectionImpl {
```
Comment (tchaikov) [inline src/crimson/os/cyan_collection.h]: what's the merit of using `ObjectStore::CollectionImpl` to access `Collection` ?

Comment (liu-chunmei) [reply src/crimson/os/cyan_collection.h]: I just use the existing implementation.  that ObjectStore::CollectionImpl as parent class will not expose collection details defined by each store to osd other code.

**Comment by liu-chunmei** [comment]
@tchaikov  PR is updated , please review, Let's make it merge ASAP, have rebased the code 2 times.

**Comment by tchaikov** [comment]
> I just use the existing implementation. that ObjectStore::CollectionImpl as parent class will not expose collection details defined by each store to osd other code.

i think we are targeting BlueStore. the only reason to derive CyanStore's `Collection` from `ObjectStore::CollectionImpl`  is to consolidate the interface of BlueStore and that of FuturizedStore, so that we can program crimson-osd using a single set of interfaces without caring about what the backend of the FuturizedStore is using.

it's relatively straightforward to understand that we need an async interface for talking to bluestore. as we will have a proxy which will call into BlueStore, and will be notified when a transaction is committed/applied.

but could you help me understand how `Collection` kicks in in this picture? as the each `BlueStore::Collection` is backed by its own `OpSequencer`. and when we flush a colllection, we wait on its `qcond` which notifies the waiters when the transaction is finished. so in the seastar world, we need an async interface for `Collection::flush()`. so, i think we will need a `FuturizedCollection` (the name may vary), so inheriting from `CollectionImpl` does not necessarily help us achieve the goal, i think. 

> Let's make it merge ASAP, have rebased the code 2 times.

i am sorry for this. i just commented on your code, and incorporated the suggested change. please lemme know if you have trouble rebasing your code, and i will help on this.

**Comment by liu-chunmei** [comment]
@tchaikov hi kefu, in osd and pg code there are CollectionRef , that is the ref defined by Ceph::os::Cyan_collection( that collection is not for all stores) , and in osd and pg, they don't care what objectstore at the backend.  and that CollectionRef can't be passed to bluestore since bluestore has difference collection definition. ( your are right the concrete collection only used inside store, but we exposed one collection to osd and pg by CollectionRef), otherwise we use #ifdef to identify what collectionRef ref to,  so we need a wrapper to hide the concrete collection in each store ( I don't think we can or need define one collection for all stores).  existing Objectstore has a wrapper ObjectStore:CollectionImpl.  I don't want to change bluestore interfaces and functions  so just use that one, but Cyan_store need meet the interface requirement since osd and pg need access cyan_store and bluestore by same way.

File: src/crimson/os/cyan_store.cc [hunk 8]
```diff
@@ -123,11 +129,12 @@ store_statfs_t CyanStore::stat() const
 }
 
 seastar::future<std::vector<ghobject_t>, ghobject_t>
-CyanStore::list_objects(CollectionRef c,
+CyanStore::list_objects(CollectionHandle ch,
                         const ghobject_t& start,
                         const ghobject_t& end,
                         uint64_t limit)
 {
+  Collection *c = static_cast<Collection*>(ch.get());
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/os/cyan_store.cc]: do you mean c? oh it is used in following logger().debug.

**Comment by liu-chunmei** [comment]
@tchaikov  PR updated, has changed futurized_collection.h definition, please review.

**Comment by tchaikov** [comment]
@liu-chunmei 

couple changes

- revert the change to s/CollectionRef/CollectionHandle/. we use ref as an implication that its smart pointer. Handle is more like an opaque type.
- add `final` specifier when inheriting from `FuturizedCollection`
- add editor variables in `futurized_collection.h`
- add missing `#include` to `futurized_collection.h`, as a header file is supposed to be self-contained.
- wrap lines at 80 characters.
- remove the alien stuff from `FuturizedCollection`. they should be in a subclass of `FuturizedCollection`. and we should not include `os/ObjectStore.h` in the abstract class' header. it's implementation details instead of a part of interface.
- always use `get_cid()` for accessing `cid`. most of the time, we don't need the concrete class of `ceph::os::Collection`.
- keep the forward declaration of `Collection` (now `FuturizedCollection`) instead of include `futurized_collection.h` if possible

**Comment by tchaikov** [comment]
retest this please.

File: src/crimson/os/cyan_store.cc [hunk 9]
```diff
@@ -646,4 +646,9 @@ boost::intrusive_ptr<Collection> CyanStore::get_collection(const coll_t& cid)
   return cp->second;
 }
 
+unsigned CyanStore::get_max_attr_name_length() const
```
Comment (tchaikov) [inline src/crimson/os/cyan_store.cc]: @liu-chunmei please avoid using merge/pull, which will bring merge commit to a PR. instead i'd suggest use rebase next time. i will do rebase on behalf of you this time.

**Pull request merged**